src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -4645,7 +4675,7 @@ public function remove_class( $class_name ): bool {
 	 * @return bool|null Whether the matched tag contains the given class name, or null if not matched.
 	 */
 	public function has_class( $wanted_class ): ?bool {
-		return $this->is_virtual() ? null : parent::has_class( $wanted_class );
+		return $this->is_virtual() ? false : parent::has_class( $wanted_class );
```

Comment:
This appears to be a small bug and isn't necessary in this PR. I'd be happy to make another PR if desired. `is_virtual` would suggest we're stopped on a tag, but the tag can't have any attributes. This should return `false`.

Reply:
hm. I would consider this more of a bug in the type signature. with active format reconstruction it _is_ possible for virtual nodes to have a class, but the `null` was meant to convey exactly what you inferred: "This tag _can have no classes_" rather than "this has _has_ no classes."

Reply:
Bug introduced in #6753

Reply:
That seems different from what the tag processor does, where "not on a tag" returns `null`, "on a tag" always returns `false` with no special handling for tags with no classes.

https://github.com/WordPress/wordpress-develop/blob/830d66c55cdb5afcaa8c0137d0c0f991258565ee/src/wp-includes/html-api/class-wp-html-tag-processor.php#L1181-L1203

Reply:
> Bug introduced in #6753

I think this was a confusion, there is no bug with types and #6753 did not introduce a bug here that I'm aware of. This is simply a question of what `null` and `false` mean here as return values.

Reply:
thanks for following up on these comments. `null` has been used to mean "cannot potentially contain that attribute or any others." maybe what we need is a comment update since we created virtual nodes.

at some point, if we allow reading these this might change, as nodes created during active format reconstruction contain the attributes of their original tags. for now, though, I prefer having a distinction between "this tag does not have this class" and "it's not possible to answer if this tag has this class"